scheduler: index out of range

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


#### What happened:
Recent commit: https://github.com/kubernetes/kubernetes/pull/99582 import an issue when multiple custom plugin configured in scheduler config file,  and those plugins are default plugins too.

It was manipulating slice with index, but the slice is updated while index is not, which will cause the index out of range, see the exceptions listed below,

```
I0708 11:11:58.252251    6748 default_plugins.go:173] "Defaut plugin is explicitly re-configured and is overriding." plugin="InterPodAffinity"
panic: runtime error: slice bounds out of range [5:3]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2.mergePluginSet(0xc00069d340, 0x7, 0x7, 0x0, 0x0, 0x0, 0xc0006a9080, 0x3, 0x8, 0x0, ...)
        /go/src/k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2/default_plugins.go:177 +0x979
k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2.mergePlugins(0xc0003c2fc0, 0xc00042afc0, 0xc00067c0c0)
        /go/src/k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2/default_plugins.go:138 +0x1f3
k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2.setDefaults_KubeSchedulerProfile(0xc00015cea0)
        /go/src/k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2/defaults.go:71 +0x4d
k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2.SetDefaults_KubeSchedulerConfiguration(0xc00015b9e0)
        /go/src/k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2/defaults.go:123 +0xf3
k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2.SetObjectDefaults_KubeSchedulerConfiguration(...)
        /go/src/k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2/zz_generated.defaults.go:55
k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2.RegisterDefaults.func3(0x1ff0280, 0xc00015b9e0)
        /go/src/k8s.io/kubernetes/pkg/scheduler/apis/config/v1beta2/zz_generated.defaults.go:35 +0x3c
k8s.io/apimachinery/pkg/runtime.(*Scheme).Default(0xc0004d5880, 0x230ba38, 0xc00015b9e0)
        /go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/scheme.go:368 +0xb5
k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Decode(0xc0000d4780, 0xc00074d000, 0xff5, 0xff6, 0x0, 0x0, 0x0, 0xc0002febe0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:168 +0x438
k8s.io/kubernetes/cmd/kube-scheduler/app/options.loadConfig(0xc00074d000, 0xff5, 0xff6, 0xff5, 0xff6, 0x0)
        /go/src/k8s.io/kubernetes/cmd/kube-scheduler/app/options/configfile.go:45 +0xc2

```
#### What you expected to happen:
config file should be handled properly without such exception

#### How to reproduce it (as minimally and precisely as possible):
override multiple default plugin in scheduler's config file.

#### Anything else we need to know?:

#### Environment:
- Kubernetes version (use `kubectl version`): latest (1.22)
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): all
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:


/assign

--- a/cmd/kube-scheduler/app/options/configfile.go
+++ b/cmd/kube-scheduler/app/options/configfile.go
@@ -87,30 +87,6 @@ func LogOrWriteConfig(fileName string, cfg *config.KubeSchedulerConfiguration, c
 	}
 	cfg.Profiles = completedProfiles
 
-	if len(fileName) > 0 {
-		// Since the default component config lists all the default plugins as enabled
-		// without disabling them, we must explicitly disable all the plugins for each
-		// extension point using the "*" expression to ensure that the generated config
-		// file is usable
-		disabledPlugins := []config.Plugin{{Name: "*"}}
-		for i := range cfg.Profiles {
-			if cfg.Profiles[i].Plugins == nil {
-				continue
-			}
-			cfg.Profiles[i].Plugins.QueueSort.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.PreFilter.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.Filter.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.PostFilter.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.PreScore.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.Score.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.Reserve.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.Permit.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.PreBind.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.Bind.Disabled = disabledPlugins
-			cfg.Profiles[i].Plugins.PostBind.Disabled = disabledPlugins
-		}
-	}
-
 	buf, err := encodeConfig(cfg)
 	if err != nil {
 		return err

--- a/pkg/scheduler/apis/config/v1beta2/default_plugins.go
+++ b/pkg/scheduler/apis/config/v1beta2/default_plugins.go
@@ -156,6 +156,8 @@ type pluginIndex struct {
 func mergePluginSet(defaultPluginSet, customPluginSet v1beta2.PluginSet) v1beta2.PluginSet {
 	disabledPlugins := sets.NewString()
 	enabledCustomPlugins := make(map[string]pluginIndex)
+	// replacedPluginIndex is a set of index of plugins, which have replaced the default plugins.
+	replacedPluginIndex := sets.NewInt()
 	for _, disabledPlugin := range customPluginSet.Disabled {
 		disabledPlugins.Insert(disabledPlugin.Name)
 	}
@@ -170,16 +172,22 @@ func mergePluginSet(defaultPluginSet, customPluginSet v1beta2.PluginSet) v1beta2
 			}
 			// The default plugin is explicitly re-configured, update the default plugin accordingly.
 			if customPlugin, ok := enabledCustomPlugins[defaultEnabledPlugin.Name]; ok {
-				klog.InfoS("Defaut plugin is explicitly re-configured and is overriding.", "plugin", defaultEnabledPlugin.Name)
-				// update the default plugin in place to preserve order
+				klog.InfoS("Default plugin is explicitly re-configured; overriding", "plugin", defaultEnabledPlugin.Name)
+				// Update the default plugin in place to preserve order.
 				defaultEnabledPlugin = customPlugin.plugin
-				// kick the plugin from the enabled list of custom plugins
-				customPluginSet.Enabled = append(customPluginSet.Enabled[:customPlugin.index], customPluginSet.Enabled[customPlugin.index+1:]...)
+				replacedPluginIndex.Insert(customPlugin.index)
 			}
 			enabledPlugins = append(enabledPlugins, defaultEnabledPlugin)
 		}
 	}
 
-	enabledPlugins = append(enabledPlugins, customPluginSet.Enabled...)
+	// Append all the custom plugins which haven't replaced any default plugins.
+	// Note: duplicated custom plugins will still be appended here.
+	// If so, the instantiation of scheduler framework will detect it and abort.
+	for index, plugin := range customPluginSet.Enabled {
+		if !replacedPluginIndex.Has(index) {
+			enabledPlugins = append(enabledPlugins, plugin)
+		}
+	}
 	return v1beta2.PluginSet{Enabled: enabledPlugins}
 }

--- a/pkg/scheduler/apis/config/v1beta2/default_plugins_test.go
+++ b/pkg/scheduler/apis/config/v1beta2/default_plugins_test.go
@@ -367,6 +367,7 @@ func TestMergePlugins(t *testing.T) {
 				Filter: v1beta2.PluginSet{
 					Enabled: []v1beta2.Plugin{
 						{Name: "Plugin1", Weight: pointer.Int32Ptr(2)},
+						{Name: "Plugin3", Weight: pointer.Int32Ptr(3)},
 					},
 				},
 			},
@@ -374,6 +375,8 @@ func TestMergePlugins(t *testing.T) {
 				Filter: v1beta2.PluginSet{
 					Enabled: []v1beta2.Plugin{
 						{Name: "Plugin1"},
+						{Name: "Plugin2"},
+						{Name: "Plugin3"},
 					},
 				},
 			},
@@ -381,6 +384,8 @@ func TestMergePlugins(t *testing.T) {
 				Filter: v1beta2.PluginSet{
 					Enabled: []v1beta2.Plugin{
 						{Name: "Plugin1", Weight: pointer.Int32Ptr(2)},
+						{Name: "Plugin2"},
+						{Name: "Plugin3", Weight: pointer.Int32Ptr(3)},
 					},
 				},
 			},
@@ -391,6 +396,38 @@ func TestMergePlugins(t *testing.T) {
 				Filter: v1beta2.PluginSet{
 					Enabled: []v1beta2.Plugin{
 						{Name: "Plugin2", Weight: pointer.Int32Ptr(2)},
+						{Name: "Plugin1", Weight: pointer.Int32Ptr(1)},
+					},
+				},
+			},
+			defaultPlugins: &v1beta2.Plugins{
+				Filter: v1beta2.PluginSet{
+					Enabled: []v1beta2.Plugin{
+						{Name: "Plugin1"},
+						{Name: "Plugin2"},
+						{Name: "Plugin3"},
+					},
+				},
+			},
+			expectedPlugins: &v1beta2.Plugins{
+				Filter: v1beta2.PluginSet{
+					Enabled: []v1beta2.Plugin{
+						{Name: "Plugin1", Weight: pointer.Int32Ptr(1)},
+						{Name: "Plugin2", Weight: pointer.Int32Ptr(2)},
+						{Name: "Plugin3"},
+					},
+				},
+			},
+		},
+		{
+			name: "RepeatedCustomPlugin",
+			customPlugins: &v1beta2.Plugins{
+				Filter: v1beta2.PluginSet{
+					Enabled: []v1beta2.Plugin{
+						{Name: "Plugin1"},
+						{Name: "Plugin2", Weight: pointer.Int32Ptr(2)},
+						{Name: "Plugin3"},
+						{Name: "Plugin2", Weight: pointer.Int32Ptr(4)},
 					},
 				},
 			},
@@ -407,8 +444,9 @@ func TestMergePlugins(t *testing.T) {
 				Filter: v1beta2.PluginSet{
 					Enabled: []v1beta2.Plugin{
 						{Name: "Plugin1"},
-						{Name: "Plugin2", Weight: pointer.Int32Ptr(2)},
+						{Name: "Plugin2", Weight: pointer.Int32Ptr(4)},
 						{Name: "Plugin3"},
+						{Name: "Plugin2", Weight: pointer.Int32Ptr(2)},
 					},
 				},
 			},

